### ISSUE TITLE
Code evolutions: Executing work on large repo's

### ISSUE THREAD

**Issue by marcodejongh** [renovatebot/renovate#9279]
**What would you like to Renovate to be able to do?**
Large mono-repos get to a point where it is tough to perform large sweeping operations on the whole repo.
For example, something as simple as turning an eslint rule on can turn into a multi-week affair due to flakey tests and code going conflicted due to ongoing development.

Dependency upgrades are when you think about out also codebase evolutions, just as an eslint rule is. To solve the example of an eslint rule, we'd want a tool that behaves similarly to Renovate to manage the PR's for the change. This means:
* Automatically rebases the branch when it goes conflicted. 
* Can potentially auto-merge
* Re-runs flakey tests (I don't believe renovate could do this right now, but it could be a nice addition for dependency upgrades too).

Later on, it would also be cool if Renovate can create these PR's per package or other entity (hence I proposed #9276.

Tools Renovate should be able to run (at least for the node ecosystem):
* Eslint
* Codemods
* Shell scripts
* Custom jobs
The idea is that a "code evolution job" should be fully automated and repeatable, similar to how a dependency upgrade renovate currently looks at is repeatable.

I'd imagine a code evolution to be scheduled using a CLI, which creates a JSON that contains the operations. For example:

```random-friendly-name-json.json
{
  tasks: [
    "eslint": { "change": "no-semi: 1" }, //Pseudo-code, there's probably a better date format to use for eslint changes.
    "codemod": { "path": "reporelativepath/to/codemod" } 
  ]
}
```
As you can see it should also be possible to schedule multiple pieces of work for a single code evolution.

**Did you already have any implementation ideas?**
Maybe this would be a manager just like npm/docker etc. But instead, this manager uses the repo checkout to determine it's jobs?

I'd be interested in building this myself

**Comment by viceice** [comment]
We already have `postUpgradeTasks`, but they will only work on selfhosted renovate because of security concerns.


https://docs.renovatebot.com/configuration-options/#postupgradetasks

**Comment by marcodejongh** [comment]
Yeah I know about postUpgradeTasks, but what I'm proposing can be run independently of a dependency upgrade.

I do envision large parts of postUpgradeTasks to be re-usable for my proposal. It's essentially a different way of defining a postUpgradeTask

**Comment by rarkins** [comment]
@marcodejongh would you like to email me for an invite to our slack channel? We can discuss things like this in a bit more free form

**Comment by HonkingGoose** [comment]
Hi @rarkins and @marcodejongh,

Did you manage to discuss this on the Slack channel?
I'm not sure if this is a thing you want in Renovate bot, and if we still want it, at what priority level?
Also if you had a brilliant idea on how to make the feature, please also mention it here in the issue. 😉

**Comment by rarkins** [comment]
Yes, we are discussing on Slack, so can keep this issue as-is until we have a clear direction forward

**Comment by marcodejongh** [comment]
@HonkingGoose I think we can use `postUpgradeAction` more extensively. I'm pretty sure there will be some gaps in Renovate, but the basic functionality we need should be there. 

I'll do a spike with renovate once I finished writing up the spec. After which I'll circle back here.
